### ISSUE TITLE
cmd/subd: export a merkle hash of on-disk state

### ISSUE THREAD

**Issue by masiulaniec** [Symantec/Dominator#551]
It would be good to have a metric that holds, for example, a float32 that obtained by taking a 4-byte prefix of a SHA-512 of the merkle hash of the entire file system. Such values could be logged to time series databases, and used by monitoring systems for making sure that the managed hosts converge to the same bits. Subd already scans the file system and calculates hashes so deriving a merkle hash should not introduce much extra overhead.

**Comment by rgooch** [comment]
I like the basic idea. Does it need to be a Merkle tree hash? That would require storing hashes in the directory inodes.
A more straight-forward implementation might be to have a modified hasher which hashes each hash that is computed.
Note that, either way, this would only expose a hash of all the file data. Inode metadata would not be captured.

**Comment by masiulaniec** [comment]
Agreed on all counts. I just wanted to put the basic idea in your head. I think metadata ought to be part of the hash.

**Comment by rgooch** [comment]
Metadata will be move complicated, but I agree that it's the kind of thing you'd want. Perhaps mtime data maybe not included?

**Comment by masiulaniec** [comment]
I can see wanting to exclude mtime for computed files.

**Comment by rgooch** [comment]
What about mtime for regular files?

**Comment by masiulaniec** [comment]
For regular files, mtime is image-defined and enforced just like any other attribute. I would include it.

**Comment by rgooch** [comment]
If mtimes for regular files are included in the hash, they will also be included for computed files, because as far as the sub is concerned, they are just regular files. It's only the **Dominator** that knows that they are computed files.

**Comment by masiulaniec** [comment]
Ack. So I don't see a reason to exclude mtime from the hash. We plan to do horizontal checks (host vs. host) and vertical (host vs. image).

**Comment by masiulaniec** [comment]
The computed files will all have equal mtime thanks to `os.Chtimes`, no?

**Comment by rgooch** [comment]
The mtime for computed files is taken from the current time when the **Dominator** sees that the computed file contents need to be changed. So, in practice, every sub is going to have a different mtime for a particular computed file. There is no horizontal consistency.

**Comment by masiulaniec** [comment]
I can see two options: a) set the mtime anyway (I realize this could confuse tools such as rsync), b) present the hasher with zero mtime for computed files.

**Comment by masiulaniec** [comment]
I understand option b) would require dominator to start revealing to subd that certain files are computed, a classification detail that is currently beautifully hidden.

**Comment by rgooch** [comment]
Yes, excluding mtime from the hash seems the best for now. I'm reluctant to complicate subd unless it's essential.

**Comment by masiulaniec** [comment]
Alternatively, the metric could be emitted at the level of the dominator server where the distinction between regular and computed files can still be made.

**Comment by rgooch** [comment]
Hm. Maybe we should take a step back at look at the problem you're trying to solve? Do you want to ensure that all machines converge to the required state and have alerting for machines which do not converge (after N attempts, say)? If that's what you're looking for, then the **Dominator** already knows this. It's currently presented in the dashboard and it could be exposed via metrics too.
